vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+have the same modify-in-place methods like `append()`. They are
+typically constructed using bare `()`, but parentheses are not strictly
+required, and you may see an implicit tuple being defined just from a
+comma separated series of expressions. Because parentheses can also be
+used to group expressions to define order of operations in expressions
+like `(x + 3) * 4`, a special syntax is required to define tuples of
+length 1: a trailing comma. Tuples are most commonly encountered in
+functions that take a variable number of arguments.
+
+```{python}
+x = (1, 2) # tuple of length 2
+type(x)
+len(x)
+x
+
+x = (1,) # tuple of length 1
+type(x)
+len(x)
+x
+
+x = () # tuple of length 0
+print(f"{type(x) = }; {len(x) = }; {x = }")
+# example of an interpolated string literals
+
+x = 1, 2 # also a tuple
+type(x)
+len(x)
+
+x = 1, # beware a single trailing comma! This is a tuple!
+type(x)
+len(x)
+```
+
+##### Packing and Unpacking
+
+Tuples are the container that powers the *packing* and *unpacking*
+semantics in Python. Python provides the convenience of allowing you to
+assign multiple symbols in one expression.
+
+For example:
+
+```{python}
+x = (1, 2, 3)
+a, b, c = x
+a
+b
+c
+```
+
+(You can access similar unpacking behavior from R using
+`` zeallot::`%<-%` ``).
+
+Tuple unpacking can occur in multiple contexts, such as iteration:
+
+```{python}
+xx = (("a", 1),
+      ("b", 2))
+for x1, x2 in xx:
+  print("x1 = ", x1)
+  print("x2 = ", x2)
+
+```
+
+If you attempt to unpack a container to the wrong number of symbols,
+Python raises an error:
+
+```{python, error = TRUE}
+x = (1, 2, 3)
+a, b, c = x # success
+a, b = x    # error, x has too many values to unpack
+a, b, c, d = x # error, x has not enough values to unpack
+```
+
+It is possible to unpack a variable number of arguments, using `*` as a
+prefix to a symbol. (You'll see the `*` prefix again when we talk about
+functions)
+
+```{python}
+x = (1, 2, 3)
+a, *the_rest = x
+a
+the_rest
+```
+
+#### Dictionaries
+
+Dictionaries are most similar to R's named lists. They are a container
+where you can retrieve items by name, though in Python the name (called
+a *key* in Python's parlance) does not need to be a string like in R. It
+can be any Python object with a `hash()` method (meaning, it can be
+almost any Python object). They can be created using syntax like
+`{key: value}`. Like Python lists, they are modified in place.
+
+```{python}
+d = {"key1": 1,
+     "key2": 2}
+d2 = d
+d
+d["key1"]
+d["key3"] = 3
+d2 # modified in place!
+```
+
+Unlike R's named lists, you cannot index into a dictionary with an
+integer to get an item at a specific index position. Dictionaries are
+*unordered* containers. (However---beginning with Python 3.7,
+dictionaries do preserve the item insertion order).
+
+```{python, error = TRUE}
+d = {"key1": 1, "key2": 2}
+d[1] # error
+```
+
+A container that closest matches the semantics of R's named list is the
```

Comment:
It was much more common in Python 2 code, but since that change in Py 3.7, it seems to be waning in favor. I added a link to https://docs.python.org/3/library/collections.html#collections.OrderedDict, for the motivated readers :)